File: src/main/java/minic/ui/MiniCAstTreeLine.java
package minic.ui;

import minic.uiapi.UiSourceSpanDto;

import java.util.Objects;

/**
 * AST 树视图中的一行。
 *
 * @param label 展示文本
 * @param depth 树深度
 * @param active 是否高亮
 * @param range 源码范围；不存在时为 {@code null}
 */
public record MiniCAstTreeLine(String label, int depth, boolean active, UiSourceSpanDto range) {
    public MiniCAstTreeLine {
        Objects.requireNonNull(label, "label");
        if (depth < 0) {
            throw new IllegalArgumentException("depth must not be negative");
        }
    }
}

File: uiweb/src/visual/MiniCAstTreeLine.ts
import type { JavaMirrorFile, UiSourceSpanDto } from "../translation/javaMirror";

export const miniCAstTreeLineMirror = {
  javaPath: "src/main/java/minic/uilocal/visual/MiniCAstTreeLine.java",
  webPath: "uiweb/src/visual/MiniCAstTreeLine.ts",
  packageName: "minic.uilocal.visual",
  exportName: "MiniCAstTreeLine",
  kind: "record",
  imports: [
    "java.util.Objects",
    "minic.uiapi.UiSourceSpanDto"
  ],
  fields: [],
  methods: [],
} as const satisfies JavaMirrorFile;

export class MiniCAstTreeLine {
  static readonly mirror = miniCAstTreeLineMirror;

  constructor(
    readonly label: string,
    readonly depth: number,
    readonly active: boolean,
    readonly range: UiSourceSpanDto | null,
  ) {
    if (depth < 0) {
      throw new Error("depth must not be negative");
    }
  }
}

export default MiniCAstTreeLine;
